### PR TITLE
IBP-4794 Manager Metadata (Name Types)

### PR DESCRIPTION
**Pull request by cuenyad**
Please Review.

Included:
https://github.com/IntegratedBreedingPlatform/Middleware/pull/1101
https://github.com/IntegratedBreedingPlatform/Workbench/pull/519

### CODE REVIEW

**Review by clarysabel — CHANGES_REQUESTED** [review]
Some comments left in the PR

File: src/main/java/org/ibp/api/java/impl/middleware/name/validator/GermplasmNameTypeValidator.java [hunk 1]
```diff
@@ -53,31 +84,80 @@ public void validate(final GermplasmNameTypeRequestDTO germplasmNameTypeRequestD
...
+		final boolean isNameTypeAssociatedToList =
+			this.germplasmNameTypeService.existNameTypeUsedInListDataProp(name);
+
+		if (isNameTypeAssociatedToList) {
+			this.errors.reject("germplasm.name.type.asociated.to.list", "");
+			throw new ApiRequestValidationException(this.errors.getAllErrors());
+		}
+	}
+
+	public void deletable(final Integer nameTypeId) {
```
[the remark this answers is not in the document]
Comment (cuenyad) [reply src/main/java/org/ibp/api/java/impl/middleware/name/validator/GermplasmNameTypeValidator.java:159]: done.

File: src/main/java/org/ibp/api/java/impl/middleware/name/validator/GermplasmNameTypeValidator.java [hunk 2]
```diff
@@ -1,48 +1,79 @@
...
 import java.util.List;
+import java.util.Optional;
 import java.util.Set;
 
 @Component
 public class GermplasmNameTypeValidator {
 
 	public static final Integer CODE_OR_NAME_MAX_LENGTH = 50;
 	public static final Integer DESCRIPTION_MAX_LENGTH = 255;
+	public static final List<String> NAME_TYPES_USED_FOR_THE_SYSTEM = Collections.unmodifiableList(
```
[the remark this answers is not in the document]
Comment (cuenyad) [reply src/main/java/org/ibp/api/java/impl/middleware/name/validator/GermplasmNameTypeValidator.java]: done.

File: src/main/java/org/ibp/api/java/impl/middleware/name/validator/GermplasmNameTypeValidator.java [hunk 3]
```diff
@@ -53,31 +84,80 @@ public void validate(final GermplasmNameTypeRequestDTO germplasmNameTypeRequestD
...
 		}
 	}
 
-	protected void validateIfCodeAlreadyExists(final String code) {
+	protected void validateIfCodeAlreadyExists(final String code, final Integer nameTypeId) {
 		final Set<String> codes = new HashSet<>(Arrays.asList(code));
 		final List<GermplasmNameTypeDTO> nameTypeDTOS = this.germplasmNameTypeService.filterGermplasmNameTypes(codes);
-		if (nameTypeDTOS != null && !nameTypeDTOS.isEmpty()) {
+
+		if (nameTypeDTOS.size() > 1 || nameTypeDTOS.size() == 1 && nameTypeId != null && !nameTypeDTOS.get(0).getId().equals(nameTypeId)) {
```
[the remark this answers is not in the document]
Comment (cuenyad) [reply src/main/java/org/ibp/api/java/impl/middleware/name/validator/GermplasmNameTypeValidator.java]: done.

File: src/test/java/org/ibp/api/java/impl/middleware/name/GermplasmNameTypeValidatorTest.java [hunk 4]
```diff
@@ -1,148 +1,212 @@
...
 		try {
-			this.germplasmNameTypeValidator.validate(germplasmNameTypeRequestDTO);
+			this.germplasmNameTypeValidator.validate(germplasmNameTypeRequestDTO, null);
 		} catch (final ApiRequestValidationException e) {
 			assertThat(Arrays.asList(e.getErrors().get(0).getCodes()), hasItem("germplasm.name.type.description.length.invalid"));
 		}
 	}
 
+	@Test
+	public void ValidateDeletable_ThrowsException_WhenNameTypeIdIsInvalid() {
```
Comment (clarysabel) [inline src/test/java/org/ibp/api/java/impl/middleware/name/GermplasmNameTypeValidatorTest.java:152]: replace in all cases Validate by testValidate @cuenyad

**Pull request merged**